packages/lit-html/src/lit-html.ts

```diff
@@ -319,15 +322,157 @@ const walker = d.createTreeWalker(
+            // this regex at the end of the tag
+            rawTextEndRegex = new RegExp(`</${match[TAG_NAME]}`, 'g');
+          }
+          regex = tagEndRegex;
+        } else if (match[DYNAMIC_TAG_NAME] !== undefined) {
+          // dynamic tag name
+          regex = tagEndRegex;
+        }
+      } else if (regex === tagEndRegex) {
+        if (match[ENTIRE_MATCH] === '>') {
+          // End of a tag. If we had started a raw-text element, use that
+          // regex
+          regex = rawTextEndRegex ?? textEndRegex;
+          // We may be ending an unquoted attribute value, so make sure we
+          // clear any pending attrNameEndIndex
+          attrNameEndIndex = -1;
+        } else {
+          attrNameEndIndex = regex.lastIndex - match[SPACES_AND_EQUALS].length;
+          attrName = match[ATTRIBUTE_NAME];
+          regex =
+            match[QUOTE_CHAR] === undefined
+              ? tagEndRegex
+              : match[QUOTE_CHAR] === '"'
+              ? doubleQuoteAttrEndRegex
+              : singleQuoteAttrEndRegex;
+        }
+      } else if (
+        regex === doubleQuoteAttrEndRegex ||
+        regex === singleQuoteAttrEndRegex
+      ) {
+        regex = tagEndRegex;
+      } else if (regex === commentEndRegex || regex === comment2EndRegex) {
+        regex = textEndRegex;
+      } else {
+        // Not one of the five state regexes, so it must be the dynamically
+        // created raw text regex and we're at the close of that element.
+        regex = tagEndRegex;
+        rawTextEndRegex = undefined;
+      }
+    }
+
+    if (DEV_MODE) {
+      // If we have a attrNameEndIndex, which indicates that we should
+      // rewrite the attribute name, assert that we're in a valid attribute
+      // position - either in a tag, or a quoted attribute value.
+      console.assert(
+        attrNameEndIndex === -1 ||
+          regex === tagEndRegex ||
+          regex === singleQuoteAttrEndRegex ||
+          regex === doubleQuoteAttrEndRegex,
+        'unexpected parse state B'
+      );
+    }
+
+    // If we're in text position, and not in a raw text element
+    // (regex === textEndRegex), we insert a comment marker. Otherwise, we
+    // insert a plain maker. If we have a attrNameEndIndex, it means we need
+    // to rewrite the attribute name to add a bound attribute suffix.
+    html +=
+      regex === textEndRegex
+        ? s + nodeMarker
+        : (attrNameEndIndex !== -1
+            ? (attrNames.push(attrName!),
+              s.slice(0, attrNameEndIndex) +
+                boundAttributeSuffix +
+                s.slice(attrNameEndIndex))
+            : s) + marker;
+  }
+  // TODO (justinfagnani): if regex is not textRegex log a warning for a
+  // malformed template in dev mode.
+  return {
+    // We don't technically need to close the SVG tag since the parser
+    // will handle it for us, but the SSR parser doesn't like that
+    html: html + strings[l] + (type === SVG_RESULT ? '</svg>' : ''),
```

Comment:
Right now we don't have a great way to minify object/class fields that are used in SSR (just top-level exports via the `$private` export). Since this is a very private API, seems chill enough to return an array (`[html, attrNames]`), so I'll do that.